[John Benedict] iP

johnbenedictyan opened the pull request:
# Aladdin Services

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Aladdin frees your mind from having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](https://google.com/).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :tada:

And it is **FREE**!

Features:
- [x]  Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

johnbenedictyan committed:
Level 1. Greet, Echo, Exit
Implement a skeletal version of Duke that starts by greeting the user, simply echos commands entered by the user, and exits when the user types bye.

johnbenedictyan committed:
Level 2. Add, List
Add the ability to store whatever text entered by the user and display them back to the user when requested.

johnbenedictyan committed:
Level 3. Mark as Done
Add the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

johnbenedictyan committed:
Level 4. ToDos, Events, Deadlines
Add support for tracking three types of tasks:
1. ToDos: tasks without any date/time attached to it e.g., visit new theme park
2. Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
3. Events: tasks that start at a specific time and ends at a specific time e.g., team project meeting on 2/10/2019 2-4pm

johnbenedictyan committed:
Test using the I/O redirection technique
Use the input/output redirection technique to semi-automate the testing of Duke

johnbenedictyan committed:
Level 5. Handle Errors
Teach Duke to deal with errors such as incorrect inputs entered by the user

johnbenedictyan committed:
A-Exceptions
Use Exceptions to handle errors

johnbenedictyan committed:
Level 6. Delete
Add support for deleting tasks from the list

johnbenedictyan committed:
A-Enums
Use Enumerations

johnbenedictyan committed:
Refactor greeting functions

johnbenedictyan committed:
Refactor task control flow with new task controller class

johnbenedictyan committed:
Refactor code

johnbenedictyan committed:
Refactor code

johnbenedictyan committed:
Add javadoc

johnbenedictyan committed:
Refactor folder structure

johnbenedictyan committed:
Level 7. Save
Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when Duke starts up.

johnbenedictyan committed:
Level 8. Dates and Times
Teach Duke how to understand dates and times

johnbenedictyan committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

johnbenedictyan committed:
A-Gradle
Automate project builds using Gradle

johnbenedictyan committed:
A-Packages
Divide classes into packages

johnbenedictyan committed:
A-MoreOOP
Refactor the code to extract out closely related code as classes

johnbenedictyan committed:
A-JUnit
Add JUnit tests to test the behaviour of the code

johnbenedictyan committed:
A-JUnit
Add JUnit tests to test the behaviour of the code

johnbenedictyan committed:
A-JavaDoc
Add JavaDoc comments to the code.

johnbenedictyan committed:
A-CodingStandard
Tweak the code to comply with a coding standard

johnbenedictyan committed:
Level 9. Find
Give users a way to find a task by searching for a keyword.

johnbenedictyan committed:
Merge branch-A-CodingStandard into master.

johnbenedictyan committed:
Merge branch-Level-9 into master.

johnbenedictyan committed:
A-CheckStyle
Use checkStyle to detect coding style violations.

johnbenedictyan committed:
Level 10. GUI
Add a GUI to Duke. Use the JavaFX technology to implement the GUI.

johnbenedictyan committed:
A-Varargs
Use Java varargs feature.

eugenelmj commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,168 @@
+     * The function returns the result of executing that command.
+     *
+     * @param command
+     *            Pass the command object to the run command method
```
This line of description should be in line with @param command, otherwise the indentation may look off. :)

johnbenedictyan replied:
Ah thank you, I will take note of this!

eugenelmj commented on src/main/java/duke/DukeCommands.java:
```diff
@@ -0,0 +1,50 @@
+ * The enum duke.Duke commands.
+ */
+public enum DukeCommands {
+    /**
```
Perhaps you could remove the individual JavaDocs for every type of commands to reduce the clutter here.

eugenelmj commented on src/main/java/duke/domain/Deadline.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    /**
+     * Of deadline.
```
Maybe you want to elaborate on what the description "Of deadline" here means? :)

eugenelmj commented on src/main/java/duke/domain/Task.java:
```diff
@@ -0,0 +1,179 @@
+     * @throws InvalidTaskSpecificationException
+     *             the invalid task specification exception
+     */
+    public static Task of(
```
Some of the indentations in the first part of this method; perhaps you might want to change some of it to improve the readability of the code? Otherwise, it may be mistaken as a form of "Deep Nesting".

johnbenedictyan replied:
I see I will see what I can do

eugenelmj commented on src/main/java/duke/domain/Event.java:
```diff
@@ -0,0 +1,96 @@
+        return Objects.hash(super.hashCode(), dateTime);
+    }
+
+    @Override
```
Good that you're abiding to the Boolean naming standards here! :)

johnbenedictyan replied:
Thank you!

eugenelmj commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,120 @@
+    public List<Task> load(String filePath) {
+        try {
+            ensureDataFileExist(filePath);
+            List<Task> initTask = new ArrayList<>();
```
Since this is a list of Task objects, the variable name should be "initTasks" instead. :)

eugenelmj commented on src/main/java/duke/taskList/TaskList.java:
```diff
@@ -0,0 +1,187 @@
+     *            Filter the tasks that are after a certain date
+     * @return A string containing the tasks that are after the datetime
+     */
+    public String outputTasksAfterString(LocalDateTime dateTime) {
```
Nice use of streams here. Well done!

eugenelmj reviewed (commented):
Overall, good effort to abstract out the classes and obeying the OOP principles! Just take note that the overall readability could be further improved; some indentations are not very consistent and may impede the readability. Also, the JavaDocs tends to be a little verbose; maybe you can use simpler sentences to summarise what the method is achieving. Otherwise, good effort! 😄

xhphoong commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,168 @@
+ */
+public class Duke {
+
+    private static final String defaultFilePath = "src/main/java/duke/data/tasks.txt";
```
Perhaps the naming of this constants should be in all uppercase?

johnbenedictyan replied:
Hmm I see, thank you for pointing this out

xhphoong commented on src/main/java/duke/domain/Task.java:
```diff
@@ -0,0 +1,179 @@
+            String text,
+            String dateTimeInput)
+            throws InvalidDateTimeException, InvalidTaskSpecificationException {
+        Boolean isEvent = Objects.equals(type, "E");
+        Boolean isDeadline = Objects.equals(type, "D");
+        Boolean isTodo = Objects.equals(type, "T");
+        if (isEvent || isDeadline || isTodo) {
+            if (isEvent || isDeadline) {
```
I like you done it in steps, by evaluating the boolean value first then use the evaluated value in the if statement. Avoid complicated expression and easier to read.

johnbenedictyan replied:
Thank you!

xhphoong commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,195 @@
+     * @param userInput
+     *            Get the user input
+     */
+    public BaseCommand parse(String userInput)
```
Maybe can find a way to shorten this method? eg : put some of the operation in separate method and call it in here?

johnbenedictyan replied:
This is definitely one area of my code that could use refactoring, I will take note of this in the future

xhphoong commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,195 @@
+                String deadlineTitle = deadlineArgs[0];
+                String deadline = deadlineArgs[1];
+
+                Task newDeadline = Task.of(
+                        "D",
+                        "0",
+                        deadlineTitle,
+                        deadline);
```
Perhaps put it in the same line to reduce the level of indentation to make it easier to read?

johnbenedictyan replied:
Oh, I thought it was easier to read like this but I shall take note

xhphoong reviewed (commented):
Overall, your code is easy to read. Most of your method is short and concise but there are few part where the method is quite long and deeply nested. Perhaps can consider to shorten those method by separating some operation into different method?

johnbenedictyan committed:
A-Assertions
Use assert feature (not JUnit assertions) to document important assumptions that should hold at various points in the code.

johnbenedictyan committed:
A-CodeQuality
Improve code quality

johnbenedictyan committed:
A-Streams
Use Streams

johnbenedictyan committed:
Merge pull request #3 from johnbenedictyan/branch-A-CodeQuality

A-CodeQuality

johnbenedictyan committed:
Merge pull request #2 from johnbenedictyan/branch-A-Streams

A-Streams

johnbenedictyan committed:
Deconflict merge

johnbenedictyan committed:
Merge pull request #1 from johnbenedictyan/branch-A-Assertions

A-Assertions

johnbenedictyan committed:
Add github workflow
GitHub Actions will run custom workflow every time certain project events are triggered (e.g., when a PR is updated, or the master branch is updated).

This workflow will help to automate mundane task like:
* checking the compilation of the gradle project
* checking the style of the project using checkstyle

johnbenedictyan committed:
Generate Javadocs
Generate javadocs to fulfil checkstyle

johnbenedictyan committed:
C-Sort
Add the sorting feature

johnbenedictyan committed:
Add ability to display help task
Add the ability for the user to ask for help and what commands are available

johnbenedictyan committed:
Implement Better UI
Updated the ui of the duke bot so that it is more visually pleasing to the users

johnbenedictyan committed:
Style Javadocs

johnbenedictyan committed:
Implement Duke Personality
Updated the command messages so that the duke bot appears more unique and friendly

johnbenedictyan committed:
Improve testing and parser

johnbenedictyan committed:
Update github pages configuration

johnbenedictyan committed:
Fix test indentation

johnbenedictyan committed:
Remove ui picture

johnbenedictyan committed:
Upload ui picture

johnbenedictyan committed:
Rename ui picture

johnbenedictyan committed:
Rename ui picture

johnbenedictyan committed:
Update github page config file

johnbenedictyan committed:
Add a User Guide

johnbenedictyan committed:
Fix gradle for release

johnbenedictyan committed:
Fix storage file creation and access
